src/typechecker.erl

```diff
@@ -3863,6 +3880,51 @@ add_any_types_pat({op, _, _Op, _Pat}, VEnv) ->
     %% Cannot contain variables.
     VEnv.
 
+-spec assign_types_to_vars_bound_more_than_once(Pats :: [gradualizer_type:abstract_pattern()],
+                                                VEnv :: map(),
+                                                Caps :: capture_vars | bind_vars) ->
+                                                       NewVEnv :: map().
+%% Assigns the type top() to variables occurring more than once in the list of
+%% patterns. This prevents the variables from being treated as match-all, thus
+%% affecting refinement.
+assign_types_to_vars_bound_more_than_once(Pats, VEnv, Caps) ->
```

Comment:
Giving variables the type `top()` feels a bit like a sledgehammer. I have the sense that we should be able to give them better types and refine the match-all logic to treat these variables better. But I don't have a concrete idea of how to do that.

Reply:
We known that these variables occurs in the pattern so we know tha they will be refined using `glb(Ty, top())`, which will be Ty in all cases. That's why top() is chosen here.